### ISSUE TITLE
Git and Acre

### ISSUE THREAD

**Issue by PaulMansour** [the-carlisle-group/Acre-Desktop#38]
The .acre folder all that it implies is fundamentally incompatible with Git.  More specifically, .acre folder and branching in Git are incompatible.  It would work ok if you never branched in Git, but I don't think it is worth the distinction. 

The .Acre folder does two things: performance, and local change files.  Both need to be sacrificed if Git is in play. The performance sacrifice is I think minimal. The local changes are nice, but local Git commits are somewhat analogous. Regardless, I think any attempt to combine the functionality is going to be very complicated,  error prone, and not worth it. 

Thus we need the following enhancement:

If an Acre project has a .git folder inside it, it is considered to be an **acre [Git] repository** .  In this case, when Acre opens the projects, the .acre folder, if it exists, is deleted.  The project is fixed from the APLSource folder. No change or delete files are written. On any change in the session, acre just writes out to the APLSource folder. Nothing else. Any call to ]undo ]redo or ]restore or ]purgehistory results in a "Not applicable to Git repository" message.  

This is one a project by project basis. Thus is is entirely reasonable to have some projects open that are git repositories and some that are not, all at the same time.

New projects should be created with .gitignore file, for safety, This is because if you create a project, and then "convert" it to git, you don't want the .acre folder getting into Git, even though you will be deleting it the next time you open the project.

**Comment by aplteam** [comment]
Why is it incompatible?

Branching a git project is a new acre project. In order to bring them back in line after changes in the branch you have to use git, yes, but when you do that with the branch and then download the new master branch acre can still incorporate those changes from the files in the APLCode folder while the branch is deleted (locally). Other way around is possible as well.

Of course there might be conflicts, but that is true in any case.

Sacrificing the change files would hurt me significantly. Yes, git commits serve the same purpose, but they are not done automatically while change files are, and that has proven to be a very good thing countless of times, partly surely because I know they are there so I don't need to be too cautious, but still.

**Comment by PaulMansour** [comment]
> Branching a git project is a new acre project.

Can you demo this?

**Comment by PaulMansour** [comment]
Forking perhaps, but branching?

**Comment by PaulMansour** [comment]
Maybe I need to rephrase. The .acre folder is incompatible with switching branches locally.

**Comment by e9gille** [comment]
Just for clarity, if you clone a repository from git, you get the entire history from revision 1. At any time you can checkout another revision either by sha1 hash or by tag/branch name. A checkout updates the files to reflect the state at a given point using the local files.

Doing this while acre keeps track of changes _will_ be confusing.

**Comment by PaulMansour** [comment]
I don't think branches can be analogous to projects.  I think repository is analogous to project.  It is possible to download a branch from GitHub?

**Comment by PaulMansour** [comment]
When you clone a repository, that is when you download it to your desktop to work on it, you get all the branches. I don't see anyway to extract a branch. I assume if you fork the repository, you could take a particular branch and make it the master of the new fork and have only one branch then.

**Comment by e9gille** [comment]
You can get a specific branch from git without cloning the repository:
https://github.com/<user>/<project>/archive/<branch>.zip

This would give you a copy of the source, but without git control (it is not a repository).

**Comment by PaulMansour** [comment]
HOLD THE PRESSES!

Let it never be said, that I am not an idiot.

It turns out that Acre Desktop as it stands is perfectly compatible with Git branching, with a small and I think very acceptable proviso - in fact it is not even a proviso.

Because Acre-Desktop looks to the APLSource folder as, well the source, it always reads in the current branch. In other words, branch switching is just exogenous changes to the APLSource folder, which is exactly what Acre Desktop is designed to deal with!   

The proviso is that the local Acre History is not branch aware, but I'm not even sure we want it to be. The local Acre history is a history of the item file in the folder- changes to that file/item may occur in any number of ways,  from the APL session,  from manually editing it, to dowloaded changes from another user to git switching branch context. It really matters not. Furthermore, exogenous changes are only noted by Acre on OpenProject.  If you branch switch a million times between consectutive calls  ]loadproject, the changes to the folder are not getting into local history.  

Now in the end, it may be that for Git users, or heavy Git branching users, that the local history and load optimization is unnecessary, and maybe we want to get rid of it or turn it off,  but as I see it now, it certainly is compatible and should not get in the way.

It is however, critical to have .gitignore file, and this must be in the project folder before it is turned into a repository.

**Comment by norberturkiewicz** [comment]
I would still like the default, or a setting, to completely ignore the .acre folder and always rebuild from source if .git is present.  Also eliminate the need for gitignore to set anything specific to Acre

**Comment by PaulMansour** [comment]
Would it make sense to have a setting to ignore ,acre, local history, optimzation, etc, regardless of the existence of the .git folder?

This setting might be useful for a number of reasons - using other SCM for example. I envision it to project level, and thus in the acrecongfig.txt file of the project.

I think the default  (for now) should be that acre keeps the local change history. You would have to turn it off explicitly. 

So, how about a new configuration parameter: **KeepLocalHistory**. The default is 1. You can set it to 0.

I use the term "LocalHistory"  to cover the change files AND the code stuff. Acre users know about local history, as the commands ]undo ]redo, etc, make this an explicit feature. Users have no idea about the the code file, etc., nor should they.

**Comment by PhilLast** [comment]
OK, Norbert, here it is. `KeepLocalHistory←'on'` will be the default and turning it off will be `'off'`.

If it's `'off'` the commands `]Restore, ]Undo, ]Redo, ]Changes` and `]PurgHistory` will message that they are inapplicable.

Acre knows nothing about any SCM, an SCM will know nothing about acre and neither need to. Nothing either does impacts on the other in any way that a user with a text editor couldn't do.

`]OpenProject` currently rebuilds the `.acre/` folder if it's missing which is the case with `]PurgeLocalHistory`. But it's already read the config file by then so it'll know not to rebuild it if the option is `'off'`.

It'll then be necessary to add conditions everywhere the codefile and change folder are referenced to ensure integrity.

Will be done in due course.

**Comment by aplteam** [comment]
@PhilLast Is this really implemented? If so, shouldn't it go into the default acreconfig.txt?!

**Comment by PhilLast** [comment]
No. I was describing what I intended to do then. Elsewhere I believe it was agreed that an interim solution to be implemented first was to be rid of the codefile in any case. That was implemented as 4.1.

This has moved the goalposts somewhat. `]purgehistory` no longer erases `.acre/` but merely empties `.acre/changes/`. The other extant member of `.acre/` is `sourcereg.dcf`. This holds the itemnames and mod-dates of the source files so that acre _knows_ what has been changed externally. A "bug" then arose when @norberturkiewicz erased `.acre/` so that the next `]openproject` _seeing_ that none of the source files were registered in `sourcereg` created change files for all of them. A fix merely suppresses that behaviour if the file is empty and populates it with the current names and dates.

 I remember reading comments since suggesting that changefiles can merely be ignored and purged occasionally.

In fact, I believe that if the option _is_ still required the solution is very simple with a guard at the top of each function affected.
